### PR TITLE
#954 #957 #1026 Customize Consul services creation in `Consul` service discovery provider

### LINKED ISSUE #954: Connection errors when Consul node names are not real DNS hosts
## Expected Behavior / New Feature
Ocelot.Provider.Consul   version 13.5.2

## Actual Behavior / Motivation for New Feature
I used `Ocelot.Provider.Consul`  and Consul, but when I make an API request，
the console display error 
```
info: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[0]
      User profile is available. Using 'C:\Users\Aries\AppData\Local\ASP.NET\DataProtection-Keys' as key repository and Windows DPAPI to encrypt keys at rest.
Hosting environment: Development
Content root path: F:\开放平台\网关\OPEN_GATEWAY\OPEN_GATEWAY
Now listening on: http://[::]:9000
Application started. Press Ctrl+C to shut down.
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://192.168.1.109:9000/open/values
dbug: Ocelot.Errors.Middleware.ExceptionHandlerMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: ocelot pipeline started
dbug: Ocelot.DownstreamRouteFinder.Middleware.DownstreamRouteFinderMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: Upstream url path is /open/values
dbug: Ocelot.DownstreamRouteFinder.Middleware.DownstreamRouteFinderMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: downstream templates are /api/{url}
info: Ocelot.RateLimit.Middleware.ClientRateLimitMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: EndpointRateLimiting is not enabled for /api/{url}
info: Ocelot.Authentication.Middleware.AuthenticationMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: No authentication needed for /open/values
info: Ocelot.Authorisation.Middleware.AuthorisationMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: /api/{url} route does not require user to be authorised
dbug: Ocelot.DownstreamUrlCreator.Middleware.DownstreamUrlCreatorMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: Downstream url is http://n1:9999/api/values
dbug: Ocelot.Requester.Middleware.HttpRequesterMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: IHttpRequester returned an error, setting pipeline error
warn: Ocelot.Requester.Middleware.HttpRequesterMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: Error making http request, exception: System.Net.Http.HttpRequestException: 不知道这样的主机。 ---> System.Net.Sockets.SocketException: 不知道这样的主机。
         at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
         at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
         at Ocelot.Requester.HttpClientHttpRequester.GetResponse(DownstreamContext context)
warn: Ocelot.Responder.Middleware.ResponderMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: Error Code: UnableToCompleteRequestError Message: Error making http request, exception: System.Net.Http.HttpRequestException: 不知道这样的主机。 ---> System.Net.Sockets.SocketException: 不知道这样的主机。
         at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
         at System.Threading.Tasks.ValueTask`1.get_Result()
         at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
         at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
         at Ocelot.Requester.HttpClientHttpRequester.GetResponse(DownstreamContext context) errors found in ResponderMiddleware. Setting error response for request path:/open/values, request method: GET
dbug: Ocelot.Errors.Middleware.ExceptionHandlerMiddleware[0]
      requestId: 0HLNVAS1E4EHV:00000001, previousRequestId: no previous request id, message: ocelot pipeline finished
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 3028.154ms 500
```

But when I use `Ocelot.Provider.Consul` version 13.5.0 there is no problem 

## Steps to Reproduce the Problem

  1. ocelot.json configuration:
  ```json
      {
  "Routes": [
    {
      "UseServiceDiscovery": true,
      "DownstreamPathTemplate": "/api/{url}",
      "DownstreamScheme": "http",
      "ServiceName": "OpenTestService",
      "UpstreamPathTemplate": "/open/{url}",
      "UpstreamHttpMethod": [ "POST", "GET" ]
    }
  ],
  "GlobalConfiguration": {
    "ServiceDiscoveryProvider": {
      "Host": "192.168.1.109",
      "Port": 8500,
      "Type": "Consul"
    }
  }
}
   ```
  consul server configuration:
```json
    {
  "encrypt": "7TnJPB4lKtjEcCWWjN6jSA==",
  "services": [
    {
      "id":"OPEN_TEST_01",
      "name":"OpenTestService",
      "tags":["OpenTestService"],
      "address":"192.168.1.109",
      "port":9999,
      "checks":[
        {
          "id": "OpenTest_Check",
          "name": "OpenTest_Check",
          "http": "http://192.168.1.109:9999/api/health",
          "interval": "10s",
          "tls_skip_verify": false,
          "method": "GET",
          "timeout": "1s"
        }
      ]
    }
  ]
}
```
## Specifications

  - Version: Ocelot.Provider.Consul version 13.5.2
  - Platform: windows 10
  - Subsystem:

### PR DESCRIPTION
**Pull request by raman-m**
## Fixes #954 #957 #1026 
- #954 
- #957 
- #1026 

## Proposed Changes
This PR introduces general customization options for the `Consul` provider, taking inspiration from the `Kube` provider customization in PR #2052. However, this PR offers a more flexible design as outlined below 👇 
  - It allows overriding the behavior of the `Consul` provider.
  - It enables the injection of a new `IConsulServiceBuilder` service into the `Consul` provider.
  - It permits the overriding of the `ConsulServiceBuilder` class's behavior.
  - The targeted virtual methods which solve the [954](https://github.com/ThreeMammals/Ocelot/issues/954) problem are:
    - `protected virtual ServiceHostAndPort GetServiceHostAndPort(ServiceEntry entry, Node node)`
    - `protected virtual string GetDownstreamHost(ServiceEntry entry, Node node)`

In conclusion, customization hinges on the virtual methods of the `ConsulServiceBuilder` class, allowing developers to override as needed for highly customized Consul setups in their environments.

Initially, I considered introducing JSON overrides for the `ServiceDiscoveryProvider` section in the global configuration of `ocelot.json`. However, I realized that approach was suboptimal. Writing C# code that inherits from the `ConsulServiceBuilder` class is a superior solution.

## Predecessors
- #909 
- #2052

### CODE REVIEW

**Comment by raman-m** [comment]
@ignacy130 Hello Ignacy!
I would value your code review and your thoughts on it as well.

@ggnaegi You are the primary reviewer for this PR. 😉
I believe the C# design is finalized, and I am currently working on the tests... You could begin the code review today...

File: src/Ocelot.Provider.Consul/OcelotBuilderExtensions.cs
Comment (ggnaegi) [inline src/Ocelot.Provider.Consul/OcelotBuilderExtensions.cs:12]: Since the users should be able to implement their own `IConsulServiceBuilder`, it might be better to have `AddConsul<T>` with T, the class implementing `IConsulServiceBuilder`

File: src/Ocelot.Provider.Consul/ConsulProviderFactory.cs
Comment (ggnaegi) [inline src/Ocelot.Provider.Consul/ConsulProviderFactory.cs:31]: Why not foresee a default dto for the configuration, instead of passing the arguments like that?

File: src/Ocelot.Provider.Consul/ConsulProviderFactory.cs
Comment (ggnaegi) [inline src/Ocelot.Provider.Consul/ConsulProviderFactory.cs:36]: The problem then... What is a Consul Service Provider? How should we handle cases like the Long Polling? I am preparing a PR implementing the consul long polling capabilities. Where should I place that?  I have implemented it, as it would be a standard functionality of the consul service provider, since I think that the Matroshka logic we have implemented here is not great, having an IServiceDiscoveryProvider embedded in another IServiceDiscoveryProvider...

File: src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs
Comment (ggnaegi) [inline src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs:8]: Why not `DefaultConsulServiceBuilder` ? Then we could propose, extra `LongPollingConsulServiceBuilder`...

File: src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs
Comment (ggnaegi) [inline src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs:37]: Why the virtuals everywhere?

**Review by ggnaegi — CHANGES_REQUESTED** [review]

**Comment by ggnaegi** [comment]
@raman-m Your idea is great, but I think we should rework it a bit first. We should think about polling and long polling and give the ability to the users to develop their own implementations.

File: src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs:37]: ### The quote 1
>In conclusion, customization hinges on **the virtual methods** of the `ConsulServiceBuilder` class, allowing developers to override as needed for highly customized Consul setups in their environments.

Thus, the concept is to permit developers to create a new class, inherit from the `ConsulServiceBuilder` class, and then simply override the necessary virtual methods to introduce new behavior.

### The quote 2
>  - The targeted virtual methods which solve the [954](https://github.com/ThreeMammals/Ocelot/issues/954) problem are:
>    - `protected virtual ServiceHostAndPort GetServiceHostAndPort(ServiceEntry entry, Node node)`
>    - `protected virtual string GetDownstreamHost(ServiceEntry entry, Node node)`

The author of issue #954 proposed reverting to the previous implementation that constructs a host from `ServiceEntry.Service.Address`.
To achieve this, a developer would need to override the method as follows:
```csharp
public class MyConsulServiceBuilder : ConsulServiceBuilder
{ 
    protected override string GetDownstreamHost(ServiceEntry entry, Node node) => entry.Service.Address;
}
```
Unfortunately, we cannot alter our current implementations as it would lead to breaking changes. It is preferable to provide developers with the flexibility to override as needed for their requirements.

File: src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs]: > Why not `DefaultConsulServiceBuilder`?

Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2067/commits/da741b4931d881a9224a6ce2936e157ab961a150 ✔️

Comment (raman-m) [reply src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs]: > Then we could propose, extra `LongPollingConsulServiceBuilder`...

I'm sorry, I don't understand. It seems that further refactoring might be beyond the scope of this PR for the #954 bug fix.
You can undertake more extensive refactoring of the Consul provider in subsequent PRs.

File: src/Ocelot.Provider.Consul/ConsulProviderFactory.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/ConsulProviderFactory.cs:31]: Apologies, that design isn't mine. 😮 It's Tom's creation! 😄
I considered refactoring the `ConsulProviderFactory` and `Consul` classes, but it was outside the bug's/PR scope. Therefore, I've only made changes to ensure the successful injection of the configuration into the `DefaultConsulServiceBuilder` constructor. We can add a TODO task for further refactoring in future iterations.

File: src/Ocelot.Provider.Consul/OcelotBuilderExtensions.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Consul/OcelotBuilderExtensions.cs:12]: @raman-m why not `AddConsul<T>` ?

File: src/Ocelot.Provider.Consul/ConsulProviderFactory.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/ConsulProviderFactory.cs:36]: Tom proposed the Matroshka logic. 🙈
I concur with 99% of the sentiment that the `ConsulProviderFactory` requires refactoring and should be transformed into a genuine DI-interface with a specific implementation class. The current `static` class implementation is indeed unattractive.

> What is a Consul Service Provider? How should we handle cases like the Long Polling?

At the moment, I'm not clear on this issue.
We have two classes of providers: `Consul` and `PollConsul`, with the following responsibilities:
- `Consul` encapsulates an `IConsulClient` object for integration with a Consul server instance. This provider executes HTTP requests to access the Consul registry.
- `PollConsul` encapsulates an `IServiceDiscoveryProvider` object, which is `Consul` at runtime. Therefore, it is tasked with service renewal through a triggering mechanism.

Will the `Consul` and `PollConsul` classes be retained in your envisioned future design?
Once more, the aim of this PR is to address the issue of multiple Consul setups through a virtual mechanism in `Consul` and `DefaultConsulServiceBuilder`, which I believe is an acceptable bug fix.
I also appreciate your desire to further develop the design of the Consul provider.

The Matroshka logic is slated for removal in the upcoming refactoring phases. 😋
Apologies if the changes have caused merge conflicts in your current feature branches, which I am unable to see. 😛

File: src/Ocelot.Provider.Consul/ConsulProviderFactory.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Consul/ConsulProviderFactory.cs:31]: @raman-m Ok, fine just add a TODO for that matter :-)

File: src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs]: Ok, fine, but then, why not `AddConsul<T>` ?

File: src/Ocelot.Provider.Consul/OcelotBuilderExtensions.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/OcelotBuilderExtensions.cs:12]: In the proposed changes, both classes, `Consul` and `DefaultConsulServiceBuilder`, contain virtual methods. It's highly likely that developers will not override the `Consul` virtual methods. However, developers should indeed inject their own subclass of `DefaultConsulServiceBuilder`. The question at hand is the precise generic signature: should it be `AddConsul<T>` or `AddConsul<T1, T2>`? Initially, `AddConsul<T>` seems adequate. For replacing `Consul` in DI, the standard approach is to use [Custom Providers](https://ocelot.readthedocs.io/en/latest/features/servicediscovery.html#custom-providers).

🆗 Working on adding new `AddConsul<T>`...

**Comment by raman-m** [comment]
>We should think about polling and long polling and give the ability to the users to develop their own implementations.

@ggnaegi, I understand you're expecting new method overrides for the next version of the provider with enhanced polling capabilities, correct? You'll have the opportunity to do so. 😉 
Are you suggesting that this bug fix should not be included in the current monthly release and that we need to work on it further? Please note that we now have a common release strategy where we release the library, the provider, and even the documentation together. If we release this bug fix in v23.3, the corresponding `Ocelot.Provider.Consul` v23.3 will also be released. If we further refactor `Ocelot.Provider.Consul` as you propose, it will be released in subsequent versions 23.x or 24.x. Therefore, I don't see the benefit of waiting for an ideal solution by delaying the delivery of the bug fix.

Let's proceed with releasing this bug fix and the package update in version [23.3](https://github.com/ThreeMammals/Ocelot/milestone/6), shall we?

File: src/Ocelot.Provider.Consul/ConsulProviderFactory.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/ConsulProviderFactory.cs:31]: Creating new TODOs poses no issues, but the real fun begins after this PR is merged... 🤣

**Review by RaynaldM — APPROVED** [review]

File: src/Ocelot.Provider.Consul/DefaultConsulServiceBuilder.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/DefaultConsulServiceBuilder.cs:35]: Not really my code but...OK
So, you think the URLs can start like this: 
- `HTTP://ABC`
- `HTTPS://ABC`

😄 
Is the goal to remove string interpolation in favor of string concatenation?

File: src/Ocelot.Provider.Consul/DefaultConsulServiceBuilder.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/DefaultConsulServiceBuilder.cs:62]: Good point! I'll move warning logging to `IsValid` method...

File: src/Ocelot.Provider.Consul/DefaultConsulServiceBuilder.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/DefaultConsulServiceBuilder.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2067/commits/18f44b9e8db8b4559affc615cf0341b538f9d605 ✔️

File: src/Ocelot.Provider.Consul/DefaultConsulServiceBuilder.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/DefaultConsulServiceBuilder.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2067/commits/347503a52ca665ed66c27af809bfc5a65f79e2ac ✔️

File: src/Ocelot.Provider.Consul/OcelotBuilderExtensions.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/OcelotBuilderExtensions.cs:24]: New method was added in commit https://github.com/ThreeMammals/Ocelot/pull/2067/commits/414d8dc3085d32028e4b9aebeee6c74845cb5d74 ✔️

File: src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs
Comment (raman-m) [reply src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs]: New method was added in commit https://github.com/ThreeMammals/Ocelot/pull/2067/commits/414d8dc3085d32028e4b9aebeee6c74845cb5d74 ✔️

**Comment by raman-m** [comment]
### The rest of the work
- [ ] Unit tests for `DefaultConsulServiceBuilder` and `Consul` with 100% coverage
- [ ] Acceptance test for #954 user scenario. See "**Steps to Reproduce the Problem**"
- [ ] Optional. Acceptance test for custom SD provider being inherited from `Consul`. See @ignacy130 commented [on Jan 30](https://github.com/ThreeMammals/Ocelot/issues/954#issuecomment-1916669129)
- [ ] Review SD docs. Move "[Store Configuration in Consul](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/configuration.rst#store-configuration-in-consul)" to "[Service Discovery](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/servicediscovery.rst)"

**Comment by raman-m** [comment]
@ggnaegi  Dev Complete! Docs written!
Waiting for your approval after final review... Could you review this weekend please?

**Review by ggnaegi — APPROVED** [review]

**Pull request merged**
